### DISCUSSION TITLE
コマンド実行ログをテキストファイルに保存する機能

### DISCUSSION THREAD

**Discussion by fushihara** [nyaosorg/nyagos#444] [Ideas]
ユーザーが入力したコマンド、そのコマンドの標準出力・標準エラーの内容、タイムスタンプをファイルに保存する機能が欲しいです。

目的としては、ログなので自分が何をしたかの記録があります。
何かPCが不調になった時、タイムスタンプを見ると昨日の何時何分何秒からおかしくなっているのでその時何のコマンドを入力したか確認したい というケースを想定しています。

今現在では以下のようにpromptに時刻を表示していますが、
これだとまずファイルへの保存がされないのと、ここに表示される時刻はコマンドが入力可能になった時刻で コマンドが入力された時刻ではないので満足できるものではありませんでした。
```
nyagos.prompt = function(this)
  local dateStr = os.date("%H:%M:%S");
  local title = "";
  default_prompt_value = default_prompt_value..(CYAN_FG..BRACKET_GT..dateStr..","..wd..BRACKET_LT..WRAP);
  return nyagos.default_prompt(default_prompt_value,title) + 1;
end
```

**Comment by hymkor** [comment]
> ユーザーが入力したコマンド、そのコマンドの標準出力・標準エラーの内容、タイムスタンプをファイルに保存する機能が欲しいです。

まず「コマンド」と「タイムスタンプ」については、ヒストリ機能の自動バックアップで 過去2000件保存しています。
`%APPDATA%\NYAOS_ORG\nyagos.history` を開いていただきますと

```
コマンドライン <TAB> その時のカレントディレクトリ <TAB> コマンド入力日時 <TAB> 入力されたNYAGOSのプロセスID
```

という形式で保存されているのが分かると思います。
（複数 NYAGOS を起動していると、出力が混ざりますが、プロセスIDでどのシェルで入力されたのかは区別できると思います）

で、次に標準出力と標準エラー出力ですが、常に記録するというのは考えにくいので、UNIX/Linux の script コマンドのような形がよいかと思いますが、ちょっとノウハウ的にしんどいので、実現するまでかなり時間がかかると思います。

（単なるリダイレクト機能ではなく、画面を出しつつ、ファイルにも記録するとなると（ファイル出力と端末出力でふるまいを変えるコマンドでもハングさせずに実行させるために）UNIX/Linux 的な疑似端末機能を使わねばならないのですが、そのあたりがなかなか難しく… Windows10 から同機能が OS に実装されていて、それを使うライブラリもあるようなのですが、まだきちんと使い方を理解しきれていないといったところです)

**Reply by fushihara** [reply]
ありがとうございます。そうか、入力は保存されていたのですね 確認させて頂きます！

標準出力・エラーについてですが、自分のイメージではターミナルソフトのRloginをイメージしていて、これは全て記録されます。
サイズ的にもテキストなので大したことは無いですし、何かあった時にgrep出来るので便利に使っています。
https://kmiya-culti.github.io/RLogin/faq.html#:~:text=%E9%81%B8%E6%8A%9E%E3%81%97%E3%81%BE%E3%81%99-,5.10,-%E9%80%9A%E4%BF%A1%E3%83%AD%E3%82%B0%E3%81%AE

仕組み的に難しいとの事で承知いたしました。
ありがとうございます！

**Comment by hymkor** [comment]
nyagos に内蔵させるのは今のところ難しいのですが、Linux の script(1) のように、標準出力・エラー出力を保存するコマンド `script.exe` を作ってみました。

- https://github.com/hymkor/script

ヒストリと同じファイルというわけにはいかないので別途照合作業などが必要になってしまいますが、`script.exe -c "nyagos.exe"` で子プロセスとして起動される nyagos とそこから呼び出されるコマンドの出力を全て `typescript` というテキストファイルに記録が出来ると思うので、これで、ある程度の記録は可能になると思います。

（本当は nyagos に内蔵できればよいのですが、Windows 7,8,Server2006 などがサポート外になってしまうんですよね。いつまでやるんだという感じではあるんですが…）
